rfcs/25-ui-systems-callbacks.md

```diff
@@ -0,0 +1,354 @@
+3. A standardized label for input dispatch that's used by core and community plugins, and a new `CoreStage::Input`. These are essential to help reduce system ordering headaches.
+4. \[Optional\] The ability to run one-shot systems with commands: [issue](https://github.com/bevyengine/bevy/issues/2192), [PR](https://github.com/bevyengine/bevy/pull/2234).
+
+## Drawbacks
+
+1. Callbacks have arbitrary power. If not carefully managed, this could create terrible spaghetti.
+2. Callbacks, like other commands, operate sequentially. This is problematic for high performance applications.
+3. There are several equivalent ways to achieve the same outcome. This choice is mostly dictated by ergonomics and subtle (but typically irrelevant) perf considerations.
+4. Serialization of callback components is likely to be challenging.
+
+## Rationale and alternatives
+
+### Why is it helpful to use the ECS for our UI logic?
+
+1. Familiar to Bevy programmers.
+2. Trivially integrated with game logic.
+3. Incredibly expressive.
+4. Benefits from other engine improvements and reduces maintenance burden.
+
+### Why do we need to add this as an engine feature?
+
+While very little code will need to be added *specifically* for this pattern,
+it is vitally important to demonstrate complex patterns in an opinionated way to users.
+
+This allows for the creation of a standardized, interoperable ecosystem,
+and guides users towards a sensible, performant and maintainable set of patterns when building their own user interfaces.
+
+### Why do we want callbacks?
+
+UI, much like scripting, tends to involve a large number of special-cased behaviors, in direct opposition to the natural patterns promoted by the ECS architecture.
+The callback pattern (and more generally, hooks) allows us to express this logic in a sane and maintainable fashion.
+
+Theoretically, everything that users could do with the callback pattern could be done with one-off systems.
+However, this clutters our scheduler (possibly hurting performance), reduces clarity and hurts compile times due to a huge number of one-off types.
+Furthermore, by storing specialized logic as data directly on UI entities it makes it dramatically easier to debug and reason about customized behavior.
+
+### Why don't we need a more complex reactivity model?
+
+Most reaction chains in UI are shockingly short.
+The most complex patterns are:
+
+1. Changing UI appearance en-masse: styles and themes should be used for this.
+2. Changing which UI elements are displayed (e.g. swapping tabs, pulling up a menu): this is precisely what `on_enter` and `on_exit` system sets in `States` are intended to solve.
+3. Handling layout changes: this should be done automatically in a single system (or group of systems) that runs in `CoreStage::PostUpdate`, rather than being scatter across our logic.
+
+For everything else, change detection and events should be more than adequate.
```

Comment:
`Generalizing behavior` gives the example of an ability task bar to show how this will scale to more complicated UIs that can't be defined at compile time.

I don't know if I agree with this section because the actual list of behavior I expect from a task bar is something like the following list:

1. There is not a 1:1 relationship to abilities and slots in the task bar. At lower levels, the player may have fewer abilities than slots. Slots will need to support being "unassigned". At higher levels, the player may have more abilities than slots. 

2. An ability doesn't need to be in the task bar. It should work correctly even when not assigned to a slot.

3. It's possible for the player to reassign abilities and slots, e.g. by drag'n'dropping. The same slot does not always correspond to the same ability. It may be possible for a player to assign the same ability to multiple slots, or attempting to do that may result in the first assignment being replaced by subsequent assignments.

4. The player can press a hot key associated with the slot (usually a number key 0-9), or they can click the slot. Either will activate the ability. It may be possible for the number keys to be remapped to other user-defined keys. It may be possible for the player to assign a key directly to an ability without requiring the ability to be in a task bar.

5. If the ability isn't ready, there will be some audio & visual feedback in response to trying to activate the ability.

6. There will be visual feedback when an ability is ready to use, on cooldown, and in progress (e.g. wind-up or other animation delays, or backswing on casting). This should work regardless of activation method -- by clicking the task bar, by pressing a hot key for that slot, by pressing an assigned key for the ability, or by selecting the ability out of the player's full list of abilities.

7. It may be possible to cancel out of the ability after starting it. Cancelling an ability has visual feedback.

8. The visual feedback for cooldown shows the time remaining in some intuitive way, e.g. a horizontal or radial wipe, or even just seconds counting down until usable again.

9. Mousing over a slot should show a tool tip displaying the ability description. This may also display contextual information, e.g. if it's on cooldown or if the player character has some status effect preventing the use of the ability. The tool tip does not pop up immediately but only after some time hovering the slot.

10. Right clicking a slot should bring up a context menu, possibly related to the ability.

11. The player may be allowed to maintain separate task bar configurations and swap between them (maybe with F1, F2, and so on).

12. The size of the task bar may change to only show slots that have abilities assigned to them. 

13. The size of the task bar may change because the player can expand it or shrink it in some fashion. As an example, it might normally be one row but the player can optionally make it more than one row. The player's configuration for other task bar configurations should be remembered between configuration changes.

14. The ability in the slot will determine the visual look of the slot. This is _normally_ just an icon, but in some games this may include additional visual information tied a specific ability -- e.g. an ability that power ups with subsequent kills may show, on the slot itself, the current number of kills. 

15. Sometimes the task bar will show the name of the ability under each slot. The ability name may need to be localized.

16. Sometimes a game allows the player to drag the task bar to another orientation, e.g. vertical instead of horizontal.

Reply:
Cool: this sounds like a fun challenge. I'll try mocking this out, and see how it might look. Some of this may be a bit hand-wavey (e.g. layout and animation), and I'll try to note which systems I expect would just be part of the engine (or a 3rd party crate).

Reply:
Can I add to this example widget focus / indirect UI navigation.

The task bar can also have a "focused" ability. Pressing `Tab` will set the focused ability to the *next* one in the current task bar. Pressing `Shift+Tab` will set the focused ability to the *previous* one in the task bar. Pressing `Space` will (attempt to) activate the focused ability (if one is focused). The actions done by `Tab`, `Shift+Tab` and `Space` may be able to be re-bound to other keys (or key combinations).